File: Linguagens Formais e Autômatos/Gals/src/br/univali/Semantico.java
package br.univali;

import java.util.Stack;
import java.util.function.BiFunction;

public class Semantico implements Constants {
	private Stack<Double> stack = new Stack<>();
	
    public void executeAction(int action, Token token)	throws SemanticError {
    	switch (action) {
    	case 1:
    		stack.push(Double.parseDouble(token.getLexeme()));
    		break;
    	case 2:
    		applyStackOperation((a, b) -> a + b);
    		break;
    	case 3:
    		applyStackOperation((a, b) -> a * b);
    		break;
    	case 4:
    		applyStackOperation((a, b) -> b - a);
    		break;
    	case 5:
    		applyStackOperation((a, b) -> b / a);
    		break;
    	case 6:
    		stack.push(-stack.pop());
    		break;
    	}
    }
    
    private void applyStackOperation(BiFunction<Double, Double, Double> consumer) {
    	stack.push(consumer.apply(stack.pop(), stack.pop()));
    }
    
    public Double getResult() {
    	return stack.peek();
    }
}

File: Linguagens Formais e Autômatos/Fermion/grammar/Semantico.cpp
#include "Semantico.h"
#include "Constants.h"

#include <iostream>

void Semantico::executeAction(int action, const Token *token) throw (SemanticError)
{
//    std::cout << action << "\n";
    switch (action) {
    case 0:
        stack.push(std::stof(token->getLexeme()));
        break;
    case 1:
        attribStack.push("");
        break;
    case 2:
        attribStack.top() = token->getLexeme();
        break;
    case 3:
        const std::string& var = attribStack.top();
        symbols[var] = stack.top();
        stack.pop();
        attribStack.pop();
        break;
    }
}

void Semantico::printSymbols()
{
    for (const std::pair<std::string, float> symbol : symbols) {
        std::cout << symbol.first << " := " << symbol.second << "\n";
    }
}

